/// <reference types="node"/>
import {
  create,
  env,
  Maybe,
} from 'sanctuary';

const S = create({ checkTypes: true, env });

// Maybe
// Alt âœ…
S.Nothing['fantasy-land/alt'](S.Nothing); // $ExpectType Maybe<never>
S.Just(1)['fantasy-land/alt'](S.Just(1)); // $ExpectType Maybe<number>
S.alt(S.Nothing)(S.Nothing); // $ExpectType Maybe<never>
S.alt(S.Just(1))(S.Nothing); // $ExpectType Maybe<number>
S.alt(S.Nothing)(S.Just(1)); // $ExpectType Maybe<number>
S.alt(S.Just(1))(S.Just(1)); // $ExpectType Maybe<number>
S.alt(S.Just(1))(S.Just('a')); // $ExpectError
// Alternative ğŸ› 
// Applicative ğŸ› 
S.of(S.Maybe)(4); // $ExpectType Maybe<number>
((): Maybe<number> => { // $ExpectType Maybe<number>
  const r = S.of(S.Maybe)(4); // $ExpectType Maybe<number>
  return r;
})();
// Apply ğŸ› 
// Bifunctor âŒ
// Category âŒ
// Chain ğŸ› 
// ChainRec ğŸ› 
// Comonad âŒ
// Contravariant âŒ
// Extend ğŸ› 
// Filterable ğŸ› 
// Foldable ğŸ› 
// Functor ğŸ› 
// Group âŒ
// Monad ğŸ› 
// Monoid if â€˜aâ€™ satisfies Semigroup ğŸ› 
// Ord if â€˜aâ€™ satisfies Ord ğŸ› 
S.lte(S.Just(1))(S.Just(4)); // $ExpectType boolean
// S.Just(1)["fantasy-land/lte"](S.Just(4)); // $ExpectType string
S.lte(S.Just(1))(S.Just(4)); // $ExpectType boolean
// Plus ğŸ› 
// Profunctor âŒ
// Semigroup if â€˜aâ€™ satisfies Semigroup ğŸ› 
// Semigroupoid âŒ
// Setoid if â€˜aâ€™ satisfies Setoid ğŸ› 
S.Just('a')["fantasy-land/equals"](S.Nothing);
S.equals(S.Just(1))(S.Just(4)); // $ExpectType boolean
// Showable
S.Just('a')["@@show"](); // $ExpectType string
S.show(S.Just(1)); // $ExpectType string
// Traversable ğŸ› 
// ----
// General
const x: Maybe<number> = S.Just("X"); // $ExpectError
S.Just<string>("X"); // $ExpectType Maybe<string>
